xplot/xplot.man
.TH XPLOT 1 "19 May 1994" "X Version 11"
.SH NAME
xplot - X plot filter
.SH SYNOPSIS
.B xplot
[-\fItoolkitoption\fP ...] [-option ...]
.SH DESCRIPTION
The
.I xplot 
program brings up a window for viewing pltfil(5) output, for example
from the
.I map
program.
The program works interactively, continuing to
accept input from stdin and plotting as it goes, until it reaches
an end of file, at which point it waits around for a mouse button
press to kill the window.
.SH OPTIONS
.I Xplot
accepts all of the standard X toolkit command line options, and the initial
geometry specification determines the resolution, with a default dimension
of a US Letter piece of paper in landscape orientation.
.PP
In addition to the standard X Toolkit command line arguments (see below),
.I xplot
also recognizes:
.TP 8
.B \-v
plot in vertical (or portrait) orientation.
.TP 8
.B \-t
use built-in X fonts instead of nplot fonts.
.TP 8
.B \-<num>x
horizontal scale factor for plot.
.TP 8
.B \-<num>y
vertical scale factor for plot.
.TP 8
.B \-<num>
horizontal and vertical scale factor for plot.
.PP
The following standard X Toolkit command line arguments may be used with 
.I xplot:
.TP 8
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.  
The default is \fIwhite\fP.
.TP 8
.B \-bd \fIcolor\fP
This option specifies the color to use for the border of the window.
The default is \fIblack\fP.
.TP 8
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding the window.
.TP 8
.B \-fg \fIcolor\fP
This option specifies the color to use for displaying text.  The default is 
\fIblack\fP.
.TP 8
.B \-fn \fIfont\fP
This option specifies the font to be used for displaying normal text.  The
default is \fI6x10\fP.
.TP 8
.B \-rv
This option indicates that reverse video should be simulated by swapping
the foreground and background colors.
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the preferred size and position of the window.
.TP 8
.B \-display \fIhost\fP:\fIdisplay\fP
This option specifies the X server to contact.
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.
.SH X DEFAULTS
This program uses the 
.I Label
widget in the X Toolkit.  It understands all of the core resource names and
classes as well as:
.PP
.TP 8
.B width (\fPclass\fB Width)
Specifies the plot window width.  The default width is US Letter paper
length. 
.TP 8
.B height (\fPclass\fB Height)
Specifies the window height.  The default height is US letter paper width.
.TP 8
.B reverseVideo (\fPclass\fB ReverseVideo)
Specifies that the foreground and background colors should be reversed.
.TP 8
.B fatlimit
limits the line fatness.
If not specified, this defaults to 2.
Lines fatter than 20 are considered to be errors.
.PP
The
.I xplot
class is "xplot", so various defaults may be set through the X resource
mechanism.
In particular, the default window geometry may be specified using the
"xplot.geometry" resource, and "xplot.fatlimit" similarly specifies the
limiting line width.
.PP
Recognized X actions are:
.TP 8
.B next()
which proceeds to the next plot in a multi-plot plot file (or ends the program
if the last plot in the file).
This is the default action if the return key or the "n" key is pressed in the
window.
.TP 8
.B quit()
which terminates the program.
This is the default action if one of the mouse buttons or the "q" key is
pressed.
.PP
The default translations for
.B xplot
are, therefore,
.nf
	   <Key>n:      next()
	   <Key>Return: next()
	   <Key>q:      quit()
	   <BtnDown>:   quit()
.fi
.SH "SEE ALSO"
X(1), plot(5), Athena Label widget
.SH BUGS
.I xplot
works fine. No bugs at all. Right. Actually there are several things
that ought to be changed, for example the reversed byte order in
plot(5) format (NOT MY FAULT), and maybe it will soon be extended
to the superplot format (I'll have to look over that stuff carefully
though), and probably also to return mouse position and anything
else the locals feel they need.
.PP
.SH AUTHORS
.nf
Arthur Smith (Lassp, Cornell University)
George Helffrich (U. Bristol)


xplot/xplot.c
#include "xplot.h"
#include <X11/Xaw/Label.h>
#if defined(__APPLE__)
#include <sys/termios.h>
#endif


int tflag = 0;
int rwidth, rheight;


/* This is supposed to set up the X window stuff, by creating a
   widget, in which the image is to be put in a background pixmap. */

int main(argc,argv)
int argc;
char **argv;
{
	int i, dimset = 0, narg, ppi = 85, x, y, is_scale();
	char *ptr, *file = NULL;
	Widget child;
	Arg wargs[20];
	void doplot(), Quit(), Next();
	XFontStruct *font;
	XGCValues gcv;
	static XtActionsRec actions[] = {
	  {"next", Next},
	  {"quit", Quit}
	};
	XtTranslations trans;
	static char translations[] = "#augment\n\
	   <Key>n:      next()\n\
	   <Key>Return: next()\n\
	   <Key>q:      quit()";
/*	   <BtnDown>:   quit()";   if you want to live dangerously */

#if defined(SGI) || defined(__linux__) || defined(__APPLE__)
        struct termios gp;
	tcflag_t old;
#else
	struct sgttyb gp;
	short old;
#endif
	int fd;
	int argf;

	progname = argv[0];

	rwidth = pxmax = 11.0*ppi; rheight = pymax = 8.5*ppi;
	xmin = ymin = 0; xmax = pxmax; ymax = pymax;

	toplevel = XtInitialize(argv[0], "xplot",NULL, 0, &argc, argv); 
	dpy = XtDisplay(toplevel);
	XtAddActions(actions,XtNumber(actions));
	trans = XtParseTranslationTable(translations);

	for(i=1; i<argc; i++) {
	   /* Parse remaining args for parameters */
	   ptr = argv[i];
	   if( is_scale(ptr) )
	      continue;
	   if (*ptr == '-') {
	      switch(*++ptr) {
		 case 'f': /* -f -- increase fatbase */
		    fatbase = atoi(++ptr);
		    break;
		 case 'v': /* -v -- vertical orientation */
		    xmax = pymax; ymax = pxmax;
		    rwidth = pymax; rheight = pxmax;
		    {int swap = pxmax; pxmax = pymax; pymax = swap;}
		    break;
		 case 't': /* -t -- Use built-in fonts */
		    tflag = 1;
		    break;
		 default:
		    {  char error_msg[64];
		       sprintf(error_msg, "Unknown option -%s.", ptr);
		       error(error_msg);
		    }
	      }
	   } else {
	      /* Must be file name to use */
	      file = ptr;
	   }
        }

	/* Open file as standard input and then condition it to work
	   unbuffered */
        instrm = stdin;
	if (file != NULL) 
	   if((instrm = freopen(file,"r", stdin)) == NULL) {
	      (void) fflush(stdout);
	      fprintf(stderr,"%s:  cannot open %s\n",progname,file);
	      exit(1);
	   }

	fd = fileno(instrm);

#if defined(SGI) || defined(__linux__) || defined(__APPLE__)
        tcgetattr(fd,&gp);
	old = gp.c_lflag;
	gp.c_lflag |= (ICANON);
	tcsetattr(fd,TCSANOW,&gp);
#else
	gtty(fd,&gp);
	old = gp.sg_flags;
	gp.sg_flags |= (CBREAK);
	stty(fd,&gp);
#endif

	argf = fcntl(fd,F_GETFL);
	fcntl(fd,F_SETFL, argf|O_NDELAY);


	/* Parse and set up fatness limit */
	if ((ptr = XGetDefault(dpy,"xplot","fatlimit"))){
	   fatlim = atoi(ptr);
	   fatlim = fatlim < 0 ? 0 : fatlim;
	}

	/* Parse and set up window geometry */
	if ((ptr = XGetDefault(dpy,"xplot","geometry"))){
	   /* Got geometry string, parse it */
	   int flags;
	   unsigned int x_w,y_w;
	   flags = XParseGeometry(ptr,&x,&y,&x_w,&y_w);
	   if (flags & WidthValue) rwidth = x_w;
	   if (flags & HeightValue) rheight = y_w;
	   dimset |= (flags & (WidthValue|HeightValue));
	   if (flags & XValue) {
	      if (XNegative & flags)
		 x += DisplayWidth(dpy,XDefaultScreen(dpy)) - rwidth;
	      dimset |= XValue;
	   }
	   if (flags & YValue) {
	      if (YNegative & flags)
		 y += DisplayHeight(dpy,XDefaultScreen(dpy)) - rheight;
	      dimset |= YValue;
	   }
	}

	pixmap = XCreatePixmap(dpy,DefaultRootWindow(dpy), rwidth, rheight,
		DefaultDepth(dpy, DefaultScreen(dpy)));

	narg = 0;
	XtSetArg(wargs[narg], XtNbitmap, pixmap); narg++;
	XtSetArg(wargs[narg], XtNinternalWidth, 0); narg++;
	XtSetArg(wargs[narg], XtNinternalHeight, 0); narg++;
        XtSetArg(wargs[narg], XtNheight, rheight); narg++;
        XtSetArg(wargs[narg], XtNwidth, rwidth); narg++;
	if (dimset & (XValue|YValue)) {
	   XtSetArg(wargs[narg], XtNx, x); narg++;
	   XtSetArg(wargs[narg], XtNy, y); narg++;
	} 

	child = XtCreateManagedWidget(progname, labelWidgetClass,
						toplevel, wargs, narg);
	XtRealizeWidget(toplevel);
	widgetwin = XtWindow(child);
	XtAugmentTranslations(child,trans);

	narg = 0;
	XtSetArg(wargs[narg], XtNbackground, &bg); narg++;
	XtSetArg(wargs[narg], XtNforeground, &fg); narg++;
	XtSetArg(wargs[narg], XtNfont, &font); narg++;
	XtGetValues(child, wargs, narg);

	gcv.foreground = fg;
	gcv.background = bg;
	gcv.font = font->fid;
	gc = XCreateGC(dpy, pixmap, GCForeground | GCBackground | GCFont,
			 &gcv);
	
	erase();

	inid = XtAddInput(fileno(instrm), (XtPointer)XtInputReadMask, doplot, NULL);
	XtMainLoop();

	/* Returned -- reset standard input */
	fcntl(fd,F_SETFL,argf);
#if defined(SGI) || defined(__linux__) || defined(__APPLE__)
	gp.c_lflag = old;
	tcsetattr(fd,TCSANOW,&gp);
#else
	gp.sg_flags = old;
	stty(fd,&gp);
#endif
        return 0;
}

void
closepl()
{
	XFreeGC(dpy,gc);
	XFreePixmap(dpy,pixmap);
	XtDestroyWidget(toplevel);
}

void
error(s)
register char *s;
{
   (void) fflush(stdout);
   fprintf(stderr,"%s: %s\n",progname,s);
   exit(1);
}

void
erase()
/* Erase clears pixmap and window */
{
   /* Clear pixmap */
   XSetForeground(dpy, gc, bg);
   XFillRectangle(dpy,pixmap,gc,0,0,rwidth,rheight);
   XSetForeground(dpy, gc, fg);
   XSetLineAttributes(dpy,gc,(unsigned)1,LineSolid,CapButt,JoinBevel);
   /* Window */
   XClearArea(dpy,widgetwin,0,0,0,0,True);
}

/* Dummy routines to keep SunOS4.1.3 libraries happy.  They are apparently
   improperly linked, leaving references to these not-really-external symbols */
__attribute__((noreturn)) void
get_wmShellWidgetClass()
{
   error("Call to get_wmShellWidgetClass"); 
}

__attribute__((noreturn)) void
get_applicationShellWidgetClass()
{
   error("Call to get_applicationShellWidgetClass"); 
}
